File: func-pipes/static/js/power2/nodes/converter.js
/*
  nodes/converter.js — Step-up / Step-down Transformer
  ─────────────────────────────────────────────────────────────────────────────
  Scales voltage up or down using a configurable target output voltage and
  efficiency factor, modelling an ideal transformer with I²R losses.

  Physics
  ───────
  turns ratio  = outVolts / _baseInVolts       (snapshotted on first live signal)
  P_in         = V_in × A_in
  P_out        = P_in × efficiency
  V_out        = V_in × turns_ratio            (tracks input — ripple passes through)
  A_out        = P_out / V_out

  _baseInVolts: the nominal input voltage captured on the first valid signal
  arrival. This locks the turns ratio so outVolts is met at that baseline.
  Changing outVolts or calling dialUp/Down re-locks to the current input.

  Extended state
  ──────────────
  outVolts     number  — target output voltage at nominal input
  step         number  — dial increment in volts
  efficiency   number  — power transfer factor (0–1; 1 = lossless)
  inVolts      number  — live measured input voltage
  inAmps       number  — live measured input current
  outAmps      number  — live computed output current
  ratio        number  — live V_out / V_in ratio
  _baseInVolts number  — snapshotted nominal input (defines turns ratio)

  States: 'off' | 'step-up' | 'step-down' | 'unity' | 'fault'
*/

class Converter extends NodeBase {

    static type  = 'converter'
    static label = 'Converter'
    static group = 'Converter'

    static catalog = [
        { key: 'conv-480v', label: 'Step-up  240→480V', outVolts: 480 },
        { key: 'conv-24v',  label: 'Step-down 240→24V', outVolts: 24  },
        { key: 'conv-12v',  label: 'Step-down 240→12V', outVolts: 12  },
        { key: 'conv-5v',   label: 'Step-down 240→5V',  outVolts: 5   },
        { key: 'psu-atx',   label: 'ATX PSU (12V)',      outVolts: 12,  efficiency: 0.88 },
    ]

    static _defaultRipple() {
        return { enabled: false, amount: 1.0, interval: 1.2 }
    }

    static defaults(id, preset = {}) {
        return {
            ...super.defaults(id, preset),
            outVolts:     preset.outVolts   ?? 120,
            step:         preset.step       ?? 10,
            efficiency:   preset.efficiency ?? 0.95,
            inVolts:      0,
            inAmps:       0,
            outAmps:      0,
            ratio:        null,
            _baseInVolts: null,
            ripple:       preset.ripple ? { ...preset.ripple } : { ...this._defaultRipple() },
        }
    }

    static configFields() {
        return [...super.configFields(), 'outVolts', 'step', 'efficiency', 'ripple']
    }

    /**
     * Transform the inbound signal. On the first live signal the turns ratio is
     * locked (snapshotting the base input voltage). Subsequent frames track any
     * ripple that passes through. Emits the converted { v, a } downstream and
     * throttles 'converter:reading' events when values change.
     * @param {Object}     panel
     * @param {Object|null} signal — upstream { v, a } or null
     * @param {PowerGraph} graph
     */
    static apply(panel, signal, graph) {
        if (!signal || signal.v <= 0 || signal.a <= 0) {
            const prev    = panel.state
            panel.state   = 'off'
            panel.inVolts = 0
            panel.inAmps  = 0
            panel.outAmps = 0
            panel.ratio   = null
            if (prev !== 'off') Converter.dispatch(panel, 'state:change', { from: prev, to: 'off' })
            graph.emit(panel, null)
            return
        }

        // Snapshot the base input voltage on first live signal.
        if (!panel._baseInVolts) panel._baseInVolts = signal.v

        const turnsRatio = panel.outVolts / panel._baseInVolts
        const pIn  = signal.v * signal.a
        const pOut = pIn * panel.efficiency
        const vOut = signal.v * turnsRatio      // tracks input; ripple passes through
        const aOut = vOut > 0 ? pOut / vOut : 0

        panel.inVolts = +signal.v.toFixed(1)
        panel.inAmps  = +signal.a.toFixed(2)
        panel.outAmps = +aOut.toFixed(2)
        panel.ratio   = +(vOut / signal.v).toFixed(3)

        if (vOut <= 0 || aOut <= 0) {
            const prev  = panel.state
            panel.state = 'fault'
            if (prev !== 'fault') Converter.dispatch(panel, 'converter:fault', { inVolts: panel.inVolts, inAmps: panel.inAmps })
            if (prev !== 'fault') Converter.dispatch(panel, 'state:change', { from: prev, to: 'fault' })
            graph.emit(panel, null)
            return
        }

        const prev  = panel.state
        panel.state = panel.ratio > 1.005
            ? 'step-up'
            : panel.ratio < 0.995
                ? 'step-down'
                : 'unity'

        if (panel.state !== prev)
            Converter.dispatch(panel, 'state:change', { from: prev, to: panel.state })

        const vOutR = +vOut.toFixed(1)
        if (vOutR !== panel._lastOutVolts || panel.outAmps !== panel._lastOutAmps) {
            panel._lastOutVolts = vOutR
            panel._lastOutAmps  = panel.outAmps
            Converter.throttle(panel, 'converter:reading', { inVolts: panel.inVolts, inAmps: panel.inAmps, outVolts: vOutR, outAmps: panel.outAmps, ratio: panel.ratio })
        }

        graph.emit(panel, { v: vOutR, a: aOut })
    }

    // ── Actions ───────────────────────────────────────────────────────────────

    static dialUp(panel, graph) {
        panel.outVolts     = +(panel.outVolts + panel.step).toFixed(1)
        panel._baseInVolts = panel.signal?.v || null
        Converter.dispatch(panel, 'converter:dial', { outVolts: panel.outVolts, direction: 'up' })
        this.apply(panel, panel.signal, graph)
        graph.updateAllGenDraws()
    }

    static dialDown(panel, graph) {
        panel.outVolts     = Math.max(1, +(panel.outVolts - panel.step).toFixed(1))
        panel._baseInVolts = panel.signal?.v || null
        Converter.dispatch(panel, 'converter:dial', { outVolts: panel.outVolts, direction: 'down' })
        this.apply(panel, panel.signal, graph)
        graph.updateAllGenDraws()
    }

    /**
     * Clamp and validate outVolts / step / efficiency after external edits (e.g.
     * config panel), re-lock the turns ratio, and re-emit downstream.
     * @param {Object}     panel
     * @param {PowerGraph} graph
     */
    static paramsChanged(panel, graph) {
        panel.outVolts   = Math.max(1,    +panel.outVolts   || 1)
        panel.step       = Math.max(0.1,  +panel.step       || 10)
        panel.efficiency = Math.min(1, Math.max(0.01, +panel.efficiency || 0.95))
        panel._baseInVolts = panel.signal?.v || null
        Converter.dispatch(panel, 'converter:params', { outVolts: panel.outVolts, efficiency: panel.efficiency })
        this.apply(panel, panel.signal, graph)
        graph.updateAllGenDraws()
    }

    /**
     * Clear all measured values and the locked turns-ratio snapshot, then
     * delegate to NodeBase.reset().
     * @param {Object}     panel
     * @param {PowerGraph} graph
     */
    static reset(panel, graph) {
        panel.inVolts      = 0
        panel.inAmps       = 0
        panel.outAmps      = 0
        panel.ratio        = null
        panel._baseInVolts = null
        panel._lastOutVolts = null
        panel._lastOutAmps  = null
        panel.powerSources = {}
        Converter.dispatch(panel, 'converter:reset', {})
        super.reset(panel, graph)
    }
}

NodeRegistry.register(Converter)

File: power_graph/src/power_graph/nodes/converter.py
"""
nodes/converter.py
──────────────────────────────────────────────────────────────────────────────
Converter — step-up / step-down transformer / power supply.

States: 'off' | 'step-up' | 'step-down' | 'unity' | 'fault'

Physics model
  turns_ratio = outVolts / _baseInVolts   (set on first live signal)
  P_out       = P_in * efficiency
  V_out       = V_in * turns_ratio
  A_out       = P_out / V_out

  If V_out would exceed maxOutVolts ('fault'), output is cut.

The dial_up() / dial_down() actions change outVolts by one step and
re-propagate through the graph.
"""

from typing import Dict, List
from power_graph.node_base import NodeBase, Signal
from power_graph.node_registry import NodeRegistry


_VOLT_STEPS = [5, 12, 24, 48, 120, 240, 480]


class Converter(NodeBase):
    """Step-up / step-down power converter."""

    type  = 'converter'
    label = 'Converter'
    group = 'Power'
    dispatch_delay = 150

    catalog = [
        {'key': 'conv-480v', 'label': 'Step-Up 480V',  'outVolts': 480, 'efficiency': 0.92},
        {'key': 'conv-240v', 'label': 'Unity 240V',    'outVolts': 240, 'efficiency': 0.98},
        {'key': 'conv-24v',  'label': 'PSU 24V',       'outVolts': 24,  'efficiency': 0.88},
        {'key': 'conv-12v',  'label': 'PSU 12V',       'outVolts': 12,  'efficiency': 0.85},
        {'key': 'conv-5v',   'label': 'PSU 5V',        'outVolts': 5,   'efficiency': 0.82},
    ]

    @classmethod
    def defaults(cls, node_id: int, preset: Dict = None) -> Dict:
        if preset is None:
            preset = {}
        base = super().defaults(node_id, preset)
        base.update({
            'outVolts':     preset.get('outVolts', 240),
            'efficiency':   preset.get('efficiency', 0.90),
            'maxOutVolts':  preset.get('maxOutVolts', 600),
            # Live readings
            'inVolts':      0.0,
            'inAmps':       0.0,
            'outAmps':      0.0,
            'ratio':        1.0,
            # Snapshot of input voltage when first energised (used as turns base)
            '_baseInVolts': None,
            'state':        'off',
        })
        return base

    @classmethod
    def config_fields(cls) -> List[str]:
        return [*super().config_fields(), 'outVolts', 'efficiency', 'maxOutVolts']

    @classmethod
    def apply(cls, panel: Dict, signal: Signal, graph):
        if signal is None:
            if panel['state'] != 'off':
                cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': 'off'})
            panel.update({
                'state': 'off', 'inVolts': 0.0, 'inAmps': 0.0,
                'outAmps': 0.0, 'ratio': 1.0, '_baseInVolts': None,
            })
            graph.emit(panel, None)
            return

        v_in      = signal.get('v', 0.0)
        a_in      = signal.get('a', 0.0)
        out_volts = panel.get('outVolts', 240)
        eff       = panel.get('efficiency', 0.90)

        # Snapshot input volts on first energising
        if panel.get('_baseInVolts') is None:
            panel['_baseInVolts'] = v_in if v_in > 0 else 1.0

        base_in  = panel['_baseInVolts']
        ratio    = out_volts / base_in
        v_out    = v_in * ratio
        p_in     = v_in * a_in
        p_out    = p_in * eff
        a_out    = (p_out / v_out) if v_out > 0 else 0.0

        panel['inVolts']  = v_in
        panel['inAmps']   = a_in
        panel['outAmps']  = a_out
        panel['ratio']    = round(ratio, 3)

        # Fault check
        if v_out > panel.get('maxOutVolts', 600):
            if panel['state'] != 'fault':
                cls.dispatch(panel, 'converter:fault', {'v_out': v_out})
            panel['state'] = 'fault'
            graph.emit(panel, None)
            return

        # Categorise state
        if abs(ratio - 1.0) < 0.02:
            new_state = 'unity'
        elif ratio > 1.0:
            new_state = 'step-up'
        else:
            new_state = 'step-down'

        if panel['state'] != new_state:
            cls.dispatch(panel, 'state:change', {'from': panel['state'], 'to': new_state})
        panel['state'] = new_state

        cls.throttle(panel, 'converter:reading',
                     {'ratio': panel['ratio'], 'v_out': round(v_out, 2), 'a_out': round(a_out, 3)})
        graph.emit(panel, {'v': round(v_out, 2), 'a': round(a_out, 3)})

    # ── Actions ───────────────────────────────────────────────────────────────

    @classmethod
    def dial_up(cls, panel: Dict, graph):
        """Increase outVolts to the next step in the standard volt ladder."""
        cur = panel.get('outVolts', 240)
        for v in _VOLT_STEPS:
            if v > cur:
                panel['outVolts']    = v
                panel['_baseInVolts'] = None  # reset turns snapshot
                cls.dispatch(panel, 'converter:dial', {'outVolts': v})
                cls._reprop(panel, graph)
                return

    @classmethod
    def dial_down(cls, panel: Dict, graph):
        """Decrease outVolts to the previous step in the voltage ladder."""
        cur = panel.get('outVolts', 240)
        for v in reversed(_VOLT_STEPS):
            if v < cur:
                panel['outVolts']    = v
                panel['_baseInVolts'] = None
                cls.dispatch(panel, 'converter:dial', {'outVolts': v})
                cls._reprop(panel, graph)
                return

    @classmethod
    def _reprop(cls, panel: Dict, graph):
        """Re-run apply() from cached sources after dial change."""
        sources  = panel.get('powerSources', {})
        combined = graph.combine_sources(sources) if sources else None
        cls.apply(panel, combined, graph)

    @classmethod
    def reset(cls, panel: Dict, graph):
        prev = panel['state']
        panel.update({'state': 'off', 'inVolts': 0.0, 'inAmps': 0.0,
                      'outAmps': 0.0, '_baseInVolts': None})
        cls.dispatch(panel, 'state:change', {'from': prev, 'to': 'off'})
        graph.emit(panel, None)


NodeRegistry.register(Converter)
